src/compiler/checker.ts

```diff
@@ -21206,18 +21236,52 @@ namespace ts {
             return true;
         }
 
+        function isReadonlyAssignmentDeclaration(d: Declaration) {
+            if (!isCallExpression(d)) {
+                return false;
+            }
+            if (!isBindableObjectDefinePropertyCall(d)) {
+                return false;
+            }
+            const objectLitType = checkExpressionCached(d.arguments[2]);
+            const valueType = getTypeOfPropertyOfType(objectLitType, "value" as __String);
+            if (valueType) {
+                const writableType = getTypeOfPropertyOfType(objectLitType, "writable" as __String);
+                if (!writableType || writableType === falseType || writableType === regularFalseType) {
+                    return true;
+                }
+                // We include this definition whereupon we walk back and check the type at the declaration because
+                // The usual definition of `Object.defineProperty` will _not_ cause literal type to be preserved in the
+                // argument types, should the type be contextualized by the call itself.
+                const writableProp = getPropertyOfType(objectLitType, "writable" as __String);
+                if (writableProp!.valueDeclaration && isPropertyAssignment(writableProp!.valueDeclaration)) {
+                    const initializer = (writableProp!.valueDeclaration as PropertyAssignment).initializer;
+                    const rawOriginalType = checkExpression(initializer);
+                    if (rawOriginalType === falseType || rawOriginalType === regularFalseType) {
+                        return true;
+                    }
+                }
+                return false;
+            }
+            const setProp = getPropertyOfType(objectLitType, "set" as __String);
+            return !setProp;
+        }
+
         function isReadonlySymbol(symbol: Symbol): boolean {
             // The following symbols are considered read-only:
             // Properties with a 'readonly' modifier
             // Variables declared with 'const'
             // Get accessors without matching set accessors
             // Enum members
+            // JS Assignments with writable false or no setter

```

Comment:
pedantically (since I spent a day thinking about terms), it's better to say `// define-property assignments with writable false or no setter`. This parallels terminology of things that are parllel, like "this-property assignments" or "prototype assignments".